### PR TITLE
add type-safe client (api and impl)

### PR DESCRIPTION
**Pull request by t1**

### CODE REVIEW

File: client-implementation/pom.xml [hunk 1]
```diff
@@ -11,12 +14,114 @@
...
+            <groupId>org.eclipse.microprofile.graphql</groupId>
+            <artifactId>microprofile-graphql-api</artifactId>
+        </dependency>
+
+        <dependency>
+            <groupId>org.projectlombok</groupId>
+            <artifactId>lombok</artifactId>
+            <version>1.18.12</version>
+            <scope>provided</scope>
+        </dependency>
```
Comment (phillip-kruger) [inline client-implementation/pom.xml:58]: Please remove Lombok (I know, it sucks)

File: client-implementation/pom.xml [hunk 2]
```diff
@@ -11,12 +14,114 @@
...
+            <artifactId>microprofile-config-api</artifactId>
+            <version>1.4</version>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>jakarta.ws.rs</groupId>
+            <artifactId>jakarta.ws.rs-api</artifactId>
+            <version>2.1.6</version>
+            <scope>provided</scope>
+        </dependency>
```
Comment (phillip-kruger) [inline client-implementation/pom.xml:70]: This might be temp (until the low level is in) right ?

File: client-implementation/pom.xml [hunk 3]
```diff
@@ -11,12 +14,114 @@
...
+        <dependency>
+            <groupId>org.projectlombok</groupId>
+            <artifactId>lombok</artifactId>
+            <version>1.18.12</version>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.eclipse.microprofile.config</groupId>
+            <artifactId>microprofile-config-api</artifactId>
+            <version>1.4</version>
```
Comment (phillip-kruger) [inline client-implementation/pom.xml:62]: Version should be in the parent pom

File: client-implementation/pom.xml [hunk 4]
```diff
@@ -11,12 +14,114 @@
...
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <version>3.8.1</version>
+                <configuration>
+                    <source>8</source>
+                    <target>8</target>
+                    <parameters>true</parameters>
```
Comment (phillip-kruger) [inline client-implementation/pom.xml:26]: Why is this needed ?

File: client-implementation/pom.xml [hunk 5]
```diff
@@ -11,12 +14,114 @@
...
+                </configuration>
+            </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-surefire-plugin</artifactId>
+                <version>2.22.2</version>
```
Comment (phillip-kruger) [inline client-implementation/pom.xml:32]: version should be in the pom.

File: client-implementation/pom.xml (hunk 2 above)
Comment (phillip-kruger) [reply client-implementation/pom.xml:70]: version might be in the parent

File: client-implementation/pom.xml [hunk 6]
```diff
@@ -11,12 +14,114 @@
...
+            <version>2.0</version>
+            <scope>provided</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter</artifactId>
+            <version>5.6.1</version>
+            <scope>test</scope>
+        </dependency>
```
Comment (phillip-kruger) [inline client-implementation/pom.xml:95]: In general - check if the version of dependencies is in the parent pom. (Or parent's parent, smallrye-parent)

File: client-implementation/pom.xml [hunk 7]
```diff
@@ -11,12 +14,114 @@
...
+            <artifactId>resteasy-client</artifactId>
+            <version>4.5.3.Final</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>ch.qos.logback</groupId>
+            <artifactId>logback-classic</artifactId>
+            <version>1.2.3</version>
+            <scope>test</scope>
+        </dependency>
```
Comment (phillip-kruger) [inline client-implementation/pom.xml:125]: Please use Jboss Logger

File: client-api/src/main/java/io/smallrye/graphql/client/typesafe/api/GraphQlClientBuilder.java [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+import java.net.URI;
+import java.util.Iterator;
+import java.util.ServiceConfigurationError;
+import java.util.ServiceLoader;
+
+public interface GraphQlClientBuilder {
+
+    static GraphQlClientBuilder newBuilder() {
+        ServiceLoader<GraphQlClientBuilder> loader = ServiceLoader.load(GraphQlClientBuilder.class);
```
Comment (phillip-kruger) [inline client-api/src/main/java/io/smallrye/graphql/client/typesafe/api/GraphQlClientBuilder.java:11]: Does RESTClient also use SPI ?

File: client-implementation/src/main/java/io/smallrye/graphql/client/typesafe/impl/cdi/GraphQlClientExtension.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+import javax.enterprise.inject.spi.AfterBeanDiscovery;
+import javax.enterprise.inject.spi.Extension;
+import javax.enterprise.inject.spi.ProcessAnnotatedType;
+import javax.enterprise.inject.spi.WithAnnotations;
+
+import io.smallrye.graphql.client.typesafe.api.GraphQlClientApi;
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class GraphQlClientExtension implements Extension {
```
Comment (phillip-kruger) [inline client-implementation/src/main/java/io/smallrye/graphql/client/typesafe/impl/cdi/GraphQlClientExtension.java:16]: Quarkus does not do CDI Extensions. We will need to find an alternative here (we can do that later)

File: client-implementation/src/main/java/io/smallrye/graphql/client/typesafe/impl/reflection/FieldInfo.java [hunk 10]
```diff
@@ -0,0 +1,56 @@
+package io.smallrye.graphql.client.typesafe.impl.reflection;
```
Comment (phillip-kruger) [inline client-implementation/src/main/java/io/smallrye/graphql/client/typesafe/impl/reflection/FieldInfo.java:1]: Where possible, use Jandex (rather that your own type). Also think about usage in Quarkus, so what can happen during build and what can not. Example, can this be scanned during build ? Then Jandex is perfect for that. (We can change that later)

File: client-implementation/pom.xml (hunk 1 above)
Comment (t1) [reply client-implementation/pom.xml]: done 😢

File: client-api/src/main/java/io/smallrye/graphql/client/typesafe/api/GraphQlClientBuilder.java (hunk 8 above)
Comment (t1) [reply client-api/src/main/java/io/smallrye/graphql/client/typesafe/api/GraphQlClientBuilder.java:11]: yes. it didn't before, but as the api is in a different module, it was necessary.

File: client-implementation/pom.xml (hunk 4 above)
Comment (t1) [reply client-implementation/pom.xml:26]: Parameter names are used as a fallback when a parameter is not annotated as `@Name`, which I really like. Too bad Oracle hasn't made this option the default.

File: client-implementation/pom.xml (hunk 3 above)
Comment (t1) [reply client-implementation/pom.xml]: done

File: client-implementation/pom.xml (hunk 2 above)
Comment (t1) [reply client-implementation/pom.xml:60]: yes

File: client-implementation/pom.xml (hunk 7 above)
Comment (t1) [reply client-implementation/pom.xml]: done

File: client-implementation/pom.xml (hunk 5 above)
Comment (t1) [reply client-implementation/pom.xml]: done

File: client-implementation/pom.xml (hunk 6 above)
Comment (t1) [reply client-implementation/pom.xml]: done

File: client-implementation/src/main/java/io/smallrye/graphql/client/typesafe/impl/cdi/GraphQlClientExtension.java (hunk 9 above)
Comment (t1) [reply client-implementation/src/main/java/io/smallrye/graphql/client/typesafe/impl/cdi/GraphQlClientExtension.java:17]: Maybe we need a Quarkus extension then?

File: client-implementation/src/main/java/io/smallrye/graphql/client/typesafe/impl/reflection/FieldInfo.java (hunk 10 above)
Comment (t1) [reply client-implementation/src/main/java/io/smallrye/graphql/client/typesafe/impl/reflection/FieldInfo.java:1]: When we add Jandex support, I had a strategy pattern in mind: if it's available, use it; otherwise use reflection. The whole reflection package hides all of this already.

File: client-implementation/pom.xml (hunk 1 above)
Comment (phillip-kruger) [reply client-implementation/pom.xml]: Thanks :)

File: client-implementation/pom.xml (hunk 4 above)
Comment (phillip-kruger) [reply client-implementation/pom.xml:25]: Jandex should give you the name. But we can add that later.

File: client-implementation/src/main/java/io/smallrye/graphql/client/typesafe/impl/cdi/GraphQlClientExtension.java (hunk 9 above)
Comment (phillip-kruger) [reply client-implementation/src/main/java/io/smallrye/graphql/client/typesafe/impl/cdi/GraphQlClientExtension.java:17]: Yes we would, let's merge like this and then see how we should change for Quarkus.

File: client-implementation/src/main/java/io/smallrye/graphql/client/typesafe/impl/reflection/FieldInfo.java (hunk 10 above)
Comment (phillip-kruger) [reply client-implementation/src/main/java/io/smallrye/graphql/client/typesafe/impl/reflection/FieldInfo.java:1]: If you do it like the server, then the Jandex stuff is in a separate module, and can be included, even in not Quarkus apps. Still easier than using reflection

Comment (phillip-kruger) [reply client-implementation/src/main/java/io/smallrye/graphql/client/typesafe/impl/reflection/FieldInfo.java:1]: We can change this after merge
